### PR TITLE
refactor: add Container better builder interfacing

### PR DESCRIPTION
**Pull request by nandor-magyar**
Refactor:
- default CLI to dyo-stable
- add Container, Builder interfaces and changed everywhere
- replace docker client.Client with its interface client.APIClient
- replace nginx image to fully qualified image names
- CLI: use data structures for repetitive tasks
- CLI: use builder hooks where applicable (Traefik file copy, migrating containers)
- CLI: One-off notification: `func NotifyOnce(...)`

Chore:
- bump golang version to 1.20
- upgrade direct dependencies
- fix gosec errors
- fix new lint errors

CI:
- add golang build cache steps to boost execution times, reduce repetition

### CODE REVIEW

File: golang/pkg/builder/container/container_builder_test.go [hunk 1]
```diff
@@ -89,19 +90,17 @@ func TestEnvPortsLabelsRestartPolicySettings(t *testing.T) {
 			"LABEL2": "1234",
 		}).
 		WithRestartPolicy(containerbuilder.AlwaysRestartPolicy).
-		WithImage("nginx:latest")
-
-	defer builderCleanup(builder)
+		WithImage("nginx:latest").
```
Comment (c3ppc3pp) [inline golang/pkg/builder/container/container_builder_test.go:93]: Please use full tag name, starting with docker.io.

File: golang/Makefile [hunk 2]
```diff
@@ -53,7 +53,7 @@ upd:
 .PHONY: down
 down:
```
Comment (c3ppc3pp) [inline golang/Makefile:54]: We talked about this, up and down should be in pairs.

**Review by c3ppc3pp — CHANGES_REQUESTED** [review]
Overall looks good, few minor changes would be appreciated.

File: golang/pkg/builder/container/container_builder_test.go (hunk 1 above)
Comment (nandor-magyar) [reply golang/pkg/builder/container/container_builder_test.go:93]: It was not added by me, but I will change it everywhere then.

File: golang/pkg/cli/container_defaults.go [hunk 3]
```diff
@@ -363,40 +366,82 @@ func GetKratos(state *State, args *ArgsFlags) *containerbuilder.DockerContainerB
...
+func getKratosInitContainer(state *State, args *ArgsFlags) containerbuilder.LifecycleFunc {
+	return func(ctx context.Context, client client.APIClient, containerName string,
+		containerId *string, mountList []mount.Mount, logger *io.StringWriter,
+	) error {
+		kratosMigrate := containerbuilder.NewDockerBuilder(context.Background()).
+			WithImage(TryImage(fmt.Sprintf("%s:%s", state.Kratos.Image, state.SettingsFile.Version), args.SpecialImageTag)).
+			WithLogWriter(nil).
+			WithName(state.Containers.KratosMigrate.Name).
+			WithoutConflict().
+			WithForcePullImage().
```
[the remark this answers is not in the document]
Comment (nandor-magyar) [reply golang/pkg/cli/container_defaults.go]: Nice catch, found another in addition to this. Removed.

File: golang/Makefile (hunk 2 above)
Comment (nandor-magyar) [reply golang/Makefile]: Done.

**Review by c3ppc3pp — APPROVED** [review]

File: golang/pkg/cli/cli.go [hunk 4]
```diff
@@ -1,12 +1,15 @@
...
 import (
+	"context"
+
 	"github.com/rs/zerolog"
 	ucli "github.com/urfave/cli/v2"
 
 	"github.com/dyrector-io/dyrectorio/golang/internal/version"
 )
 
+//nolint:funlen
```
Comment (robot9706) [inline golang/pkg/cli/cli.go:12]: 🏗

File: golang/pkg/cli/display.go [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+func PrintWelcomeMessage(settingsPath string) {
+	log.Info().Msgf("The config file is located at %s, where you can turn this message off.", settingsPath)
+	log.Info().Msgf("If you have any questions head to our Discord - https://discord.gg/pZWbd4fxga ! We're happy to help!")
+	log.Info().Msgf("You can learn more about the project at https://docs.dyrector.io, if you found this project useful please " +
+		"give us a star - https://github.com/dyrector-io/dyrectorio")
+}
+
+func PrintInfo(state *State) {
+	log.Warn().Msg("🦩🦩🦩 Use the CLI tool only for NON-PRODUCTION purpose. 🦩🦩🦩")
```
Comment (robot9706) [inline golang/pkg/cli/display.go:18]: I know you did not change this but `...NON-PRODUCTION purposes.` looks better, no?

File: golang/Makefile [hunk 6]
```diff
@@ -40,20 +40,28 @@ endif
 
 GOPARAMS:=GOOS=$(GOOS) CGO_ENABLED=0 $(GRPC_DEBUG_FLAGS)
 
+# cli start fully containerized stack, stable branch
```
Comment (robot9706) [inline golang/Makefile:43]: `down` has the `--prefix dyo-stable`, doesn't the `up` command need it too?

File: golang/pkg/cli/cli.go (hunk 4 above)
Comment (nandor-magyar) [reply golang/pkg/cli/cli.go:12]: could not be avoided :cry:

File: golang/pkg/builder/container/docker_container.go [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+	container      *types.Container
+	preStartHooks  []LifecycleFunc
+	postStartHooks []LifecycleFunc
+}
+
+func NewDockerContainer(cont *types.Container, preStartHooks, postStartHooks []LifecycleFunc) DockerContainer {
+	return DockerContainer{container: cont, preStartHooks: preStartHooks, postStartHooks: postStartHooks}
+}
+
+func (d DockerContainer) Sync() error {
```
[the remark this answers is not in the document]
Comment (nandor-magyar) [reply golang/pkg/builder/container/docker_container.go:36]: Removed.

File: golang/Makefile (hunk 6 above)
Comment (nandor-magyar) [reply golang/Makefile:43]: That is the default in case of up, same with down, it is not necessary.

File: Makefile [hunk 8]
```diff
@@ -1,28 +1,38 @@
 
 SHELL = /bin/sh
 
-## shortcut to start stack
+## shortcut to start stack, fully containerized, stable stable build
```
Comment (polaroi8d) [inline Makefile:4]: stable stable?

**Review by polaroi8d — APPROVED** [review]
(informal) LGTM

File: Makefile (hunk 8 above)
Comment (nandor-magyar) [reply Makefile]: Fixed.

**Pull request merged**